keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1210,230 +1099,251 @@ server-side-apply (SSA) simpler. Patching individual list elements is not
+//
+// Whenever a driver needs to update a pool, it bumps the pool generation number
+// and updates all slices with that new number and any new device definitions. A consumer
+// must only use device definitions from slices with the highest generation number
+// and ignore all others.
+//
+// If necessary, a consumer can check the number of total devices in a pool (included
+// in each slice) to determine whether its view of a pool is complete.
+//
+// For devices that are not local to a node, the node name is not set. Instead,
+// the driver may use a node selector to specify where the devices are available.
 type ResourceSlice struct {
     metav1.TypeMeta
     // Standard object metadata
+    // +optional
     metav1.ObjectMeta
 
-    // NodeName identifies the node which provides the resources
-    // if they are local to a node.
-    //
-    // A field selector can be used to list only ResourceSlice
-    // objects with a certain node name.
-    NodeName string
+    Spec ResourceSliceSpec
+
+    Status ResourceSliceStatus
+}
 
+type ResourceSliceSpec struct {
     // DriverName identifies the DRA driver providing the capacity information.
     // A field selector can be used to list only ResourceSlice
     // objects with a certain driver name.
+    //
+    // Must be a DNS subdomain and should end with a DNS domain owned by the
+    // vendor of the driver.
     DriverName string
 
-    ResourceModel
+    // PoolName is used to identify devices. For node-local devices, this
+    // is often the node name, but this is not required.
+    //
+    // It must not be longer than 253 and must consist of one or more DNS sub-domains
+    // separated by slashes.
+    PoolName string
+
+    // NodeName identifies the node which provides the devices.
+    // A field selector can be used to list only ResourceSlice
+    // objects belonging to a certain node.
+    //
+    // This field can be used to limit access from nodes to slices with
+    // the same node name. It also indicates to autoscalers that adding
+    // new nodes of the same type as some old node might also make new
+    // devices available.
+    //
+    // NodeName and NodeSelector are mutually exclusive.
+    // If both are unset, then the devices are available in the
+    // entire cluster.
+    //
+    // +optional
+    NodeName string
+
+    // Defines which nodes have access to the devices in the pool.
+    // Must not be set when the node name is set.
+    //
+    // NodeName and NodeSelector are mutually exclusive.
+    // If both are unset, then the devices are available in the
+    // entire cluster.
+    //
+    // +optional
+    NodeSelector *v1.NodeSelector
+
+    // The generation gets bumped in all slices of a pool whenever device
+    // definitions change. A consumer must only use device definitions from slices
+    // with the highest generation number and ignore all others.
+    PoolGeneration int64
+
+    // The total number of devices in the pool.
```

Comment:
Yes, that's a good point, and more how I was originally thinking of it too. The scheduler doesn't care if it has a complete view when it is trying to scheduler; it either succeeds or fails.

Is there a specific scenario where that's not good enough?

Reply:
The intent is that the scheduler waits at least in cases where it matters ("all devices"). See https://github.com/kubernetes/enhancements/pull/4709#discussion_r1642593833 for a more complete discussion.

We could also say "always wait". That doesn't matter much for other cases at the moment (if it knows about enough devices to satisfy a claim, adding more to choose from doesn't change much), but once we have scoring it would be good to not make a choice based on known incomplete information.

Reply:
The "all" case makes this necessary, I agree.